### ISSUE TITLE
tuning tables off for using A as defining note for the pitch

### ISSUE THREAD

**Issue by mirabilos** [billhails/MuseScore-plugins#5]
Most of the pre-existing tables, except the new Bach/Lehman ones I proposed in #4 have the constant note at C instead of A. However, MuseScore can only tune using A4 as reference, so the A in the cent table should always be exactly `900.0`, and the other cent offsets need to be transposed for that to work.

For Bach/Lehman, he even gives it as example on his website:

“If you use an app or device that allows definitions of Custom
temperaments, such as ClearTune, set it up with the following cent
offsets from equal temperament:

    C (+5.9)   C♯ (+3.9)   D (+2)     E♭ (+3.9)   E (-2)     F (+7.8)   F♯ (+2)     G (+3.9)   G♯ (+3.9)   A (0)      B♭ (+3.9)   B (0). 

[…] “Enter the offsets from C, not A. The numbers to use are:

    C (0.0)    C♯ (-2.0)   D (-3.9)   E♭ (-2.0)   E (-7.8)   F (2.0)    F♯ (-3.9)   G (-2.0)   G♯ (-2.0)   A (-5.9)   B♭ (-2.0)   B (-5.9). 

It can be observed that the bottom line is created from the top line by subtracting 5.9, with the exception of F (-5.8), which may come from rounding, but it’s close enough. (Or it may be a typo in the C pitches, I triple-checked the 7.8 as I could not get it at first either.) This 5.9 is exactly the difference between the cent deltas of the respective tuning base notes.

So, I think we need to adjust all the other tables (those I did not touch in #4) to keep A constant.

**Comment by nefariousphd** [comment]
Do you have a reference that explains that the notes need to keep A constant? I know there's limited info out there and some of it is old, but the plug-in reference for notes (https://musescore.org/en/plugin-development/note-object) makes it seem that the tuning can be applied to any note.

**Comment by mirabilos** [comment]
Hi,

>Do you have a reference that explains that the notes need to keep A
>constant? I know there's limited info out there and some of it is old,
>but the plug-in reference for notes
>(https://musescore.org/en/plugin-development/note-object) makes it
>seem that the tuning can be applied to any note.

indeed, this was more of a thing of what the user wants to hear.

Just today I’ve wondered about this as well, as I generally tune to
C4=256 Hz, or rather MuseScore’s equivalent tuning of A4=430.539 Hz.
I was wondering whether I really need to add a second set of rows for
Bach/Lehman tuning on C to the plugin (which would also require me to
add the tuning to the .mscx file, so playback everywhere matches) but
haven’t gotten around to doing so.

I think we need an algorithmic description of the temperaments, like
https://github.com/gstlaurent/ttuner has, and a second option in the
dialogue asking which tone to keep constant, defaulting to A4, but
also offering other options. This is, of course, nōn-trivial. (After
all, algorithmic is how I came to the cent delta values for my PR
related to the Bach/Lehman temperament; I generally know how to do
this, but it differs for each temperament, of course.)

I think this needs further research, while still being a valid issue.
I’ll do some more of this research and perhaps a proof-of-concept,
if you would kindly leave the issue open, in whatever passes as my
spare time nowadays (I’m working with dozens of, mostly Open Source
related, projects).

**Comment by mirabilos** [comment]
I’ve since brushed up on theory behind temperaments (and how to _properly_ calculate them…) **even more**, and am currently working on a reference implementation (in a programming language I know really well, as I’m not familiar with QML) of a rule-based system that can take rules similar to ttuner, although with an extended (more expressive), restricted (more precise / less ambiguous) rule language, and takes additionally a pitch standard (assign one frequency to one note), and will be able to calculate frequencies for all 128 MIDI notes, as well as cent deltas. (Currently, I’ll only support using MIDI notes, i.e. enharmonics get assigned the same frequency. But for the purpose of MuseScore tuning, that is enough.)

Hmm… perhaps I should also add a feature that gives the correct MuseScore `synthesizer.xml` A₄ value for any given nōn-A₄-based tuning…

I *think* I’ll also add a feature to allow retuning of `.mscx` file parts without starting MuseScore (mostly to avoid relayouting issues when making hidden staves visible to be able to use them with your plugin; this is dead easy because…

          <Note>
            <pitch>69</pitch>
            <tpc>17</tpc>
            <tuning>-5.9</tuning>
            </Note>

… the `.mscx` file format is trivial enough); this _will **not**_ obsolete your plugin however, as a MuseScore plugin is *much* easier to use for the average user, and my program will not work as-is on nōn-Unix platforms anyway. So do not fear.

I’m *also*, in parallel, working on describing *all* temperaments included with either ttuner or your plugin in a rule-based way, so they’ll be usable with all kinds of root notes. This involves quite a bit of historic research, but I’ve got most of them already.

Finally, I’ll be working on a transposition feature, so that we need no separate “Bach/Lehman Chorale” version like in https://github.com/billhails/MuseScore-plugins/pull/4 — the user can simply select “Bach/Lehman 1722” and “transpose 2 semitones up”. This will also allow one to use [Pythagorean tuning](https://en.wikipedia.org/wiki/Pythagorean_tuning#Method) with a different centre note than **D**.

Once I’ve got this working well and stable algorithms, I’ll try to translate into QML and submit another PR. This post is to keep you updated and possibly invite any feedback.

**Comment by billhails** [comment]
Thanks, I've not had much time to go through this, but it's good to know you're helping out.

**Comment by nefariousphd** [comment]
FWIW, I found a reference that states that Kirnberger 2 should be based off of C, rather than A.
"The objective in K-II is to have C major and as many “nearby” scales as possible to be just (have perfect chords), so the tuning is started from middle C (C4 = 261.6Hz...)"
https://fundamentals-of-piano-practice.readthedocs.io/en/latest/chapter2/CH2.6.html
As for the other WT, I can't find anything definitive, but I'd guess that by the way they are constructed (to give more dissonance to keys further from C), that they too were tuned starting from C.
Now, whether that would be the ET-12 C4 of 261.6 Hz or some other standard C4 frequency, I cannot tell.

**Comment by mirabilos** [comment]
nefariousphd dixit:

>FWIW, I found a reference that states that Kirnberger 2 should be

Hm, we have Kirnberger 3, not 2, in the list.

>based off of C, rather than A.

In my new setup, the root note of tuning and the note that’s kept
constant (pitch standard) are separately given, so that will not
be a problem any more. (I’ve written 27138 bytes already, and still
way more to come, but a lot of it is comments, too.)

>"The objective in K-II is to have C major and as many “nearby” scales
>as possible to be just (have perfect chords), so the tuning is started
>from middle C (C4 = 261.6Hz...)"
>https://fundamentals-of-piano-practice.readthedocs.io/en/latest/chapter2/CH2.6.html

Thanks. I’ve also found some other extremely useful references for
various scales, but I found that, before creating rule files for
all of them, I need to finish the script, so I can compare its output
in cent deltas with some of the references that only give them.

I also implemented loading of .scl (Scala files), although the user
must manually say which note is used as first note of the scale, and
for 12-tone scales only; this still gives us instant access to about
2000 tunings, although many of them are inaccurate (given in cent or
rounding fractional commas to rational fractions instead of properly
using roots).

I’m making progress every couple of days, as I come to it, be it more
coding, more comment writing, or just more research.

bye,
//mirabilos
-- 
„Cool, /usr/share/doc/mksh/examples/uhr.gz ist ja ein Grund,
mksh auf jedem System zu installieren.“
	-- XTaran auf der OpenRheinRuhr, ganz begeistert
(EN: “[…]uhr.gz is a reason to install mksh on every system.”)

**Comment by mirabilos** [comment]
Dixi quod…

>nefariousphd dixit:
>
>>FWIW, I found a reference that states that Kirnberger 2 should be
>
>Hm, we have Kirnberger 3, not 2, in the list.

Kirnberger 3 is currently defined like this:

    DESC Kirnberger Ⅲ
    
    ! Start C-E as pure major third, then pure fifths up to F♯ and down to D♭
    ! Cₙ-Eₙ₊₂ is the definition of the syntonic comma, divide it on four fifths
    ! The schisma sits between F♯ and C♯ to close the circle.
    
    A♭ = D♭
    E♭ = A♭
    B♭ = E♭
    F  = B♭
    C  = F
    G  = C - 1/4 S
    D  = G - 1/4 S
    A  = D - 1/4 S
    E  = A - 1/4 S
    B  = E
    F♯ = B
    
    VALIDATE
    
    E  = C
    C♯ = F♯ - D

In case this helps. (The syntax is from ttuner, only the `D` comma is
new, and there are additional intervals, not used here. And I insist
on having the left-hand-side be the to-be-defined note (i.e. the higher
one, here).)

**Comment by billhails** [comment]
Another update, I've added both "Root Note" and "Pure Tone" adjustments to the plugin. Please see the updated docs.

**Comment by billhails** [comment]
I think the "Pure Tone" and "Root Note" features now address this issue so I'm closing it. feel free to re-open if you don't agree.
